### PR TITLE
Fixes and error handling rework for RAUC boot selection layer

### PR DESCRIPTION
**Pull request by ejoerns**
This series targets different issues in Barebox bootchooser (boot selection layer) implementation.

The two main ones are:

* This is one of the components that still lacks appropriate GError handling, giving us only insufficient information about failures
* barebox-state backend did not properly handle disabled slots for setting slots as primary. This also additionally requires reading state info instead of only writing it

### CODE REVIEW

File: src/bootchooser.c [hunk 1]
```diff
@@ -44,27 +44,36 @@ static gboolean bootchooser_order_primay(RaucSlot *slot, GString **value) {
...
 }
 
+typedef struct {
+	guint32 prio;
+	guint32 attempts;
+} BareboxSlotState;
+
+#define BOOTSTATE_PREFIX "bootstate"
+
 #if 0
```
Comment (jluebbe) [inline src/bootchooser.c]: According to the commit message, the function should be reenabled.

File: src/bootchooser.c [hunk 2]
```diff
@@ -22,12 +22,11 @@ GQuark r_bootchooser_error_quark (void)
 #define UBOOT_FWSETENV_NAME "fw_setenv"
 #define UBOOT_FWGETENV_NAME "fw_printenv"
 
-static gboolean bootchooser_order_primay(RaucSlot *slot, GString **value) {
+static GString *bootchooser_order_primay(RaucSlot *slot) {
 	GString *order = g_string_sized_new(10);
 	GList *slots;
 
 	g_return_val_if_fail(slot, FALSE);
```
Comment (jluebbe) [inline src/bootchooser.c]: We can't return FALSE here anymore.

File: src/bootchooser.c [hunk 3]
```diff
@@ -508,9 +501,13 @@ static gboolean uboot_set_primary(RaucSlot *slot) {
 	/* Add updated slot as first entry in new boot order */
 	g_string_append(order_new, slot->bootname);
 
-	res = uboot_env_get("BOOT_ORDER", &order_current, NULL);
-	if (!res && !bootchooser_order_primay(slot, &order_current))
-		goto out;
+	res = uboot_env_get("BOOT_ORDER", &order_current, &ierror);
+	if (!res) {
+		g_message("Unable to obtain BOOT_ORDER");
```
Comment (jluebbe) [inline src/bootchooser.c]: Maybe indicate that this is not cirtical and we fall back to a default order?

File: test/install-content/custom_handler.sh [hunk 4]
```diff
@@ -24,6 +24,14 @@ done
 exit_if_empty $RAUC_MOUNT_PREFIX
 mkdir -p $RAUC_MOUNT_PREFIX/image
 
+# This is only for testing
```
Comment (jluebbe) [inline test/install-content/custom_handler.sh]: Should this change be squashed into the one changing the code?

File: src/bootchooser.c (hunk 1 above)
Comment (ejoerns) [reply src/bootchooser.c]: This would require squashing commits if we intend to keep the commit compilable.

Comment (jluebbe) [reply src/bootchooser.c]: Hmm, OK. Then explain that in the commit message.

File: src/bootchooser.c (hunk 3 above)
Comment (ejoerns) [reply src/bootchooser.c]: I assume changing it to 
``` c
g_message("Unable to obtain BOOT_ORDER, using defaults");
```

would be sufficient, as the log level indicates being non-critical?

File: test/install-content/custom_handler.sh (hunk 4 above)
Comment (ejoerns) [reply test/install-content/custom_handler.sh]: Yes, could, was more ore less a post-fix.

**Comment by ejoerns** [comment]
Updated commit messages, squashed and fixed as requested.

``` diff
diff --git a/src/bootchooser.c b/src/bootchooser.c
index bf8888b..504dfa6 100644
--- a/src/bootchooser.c
+++ b/src/bootchooser.c
@@ -26,7 +26,7 @@ static GString *bootchooser_order_primay(RaucSlot *slot) {
        GString *order = g_string_sized_new(10);
        GList *slots;
 
-       g_return_val_if_fail(slot, FALSE);
+       g_return_val_if_fail(slot, NULL);
 
        g_string_append(order, slot->bootname);
 
@@ -504,7 +504,7 @@ static gboolean uboot_set_primary(RaucSlot *slot, GError **error) {
 
        res = uboot_env_get("BOOT_ORDER", &order_current, &ierror);
        if (!res) {
-               g_message("Unable to obtain BOOT_ORDER");
+               g_message("Unable to obtain BOOT_ORDER, using defaults");
                g_clear_error(&ierror);
 
                order_current = bootchooser_order_primay(slot);
```

**Pull request merged**